### PR TITLE
Simplify GenericType

### LINKED ISSUE #16102: Simplify GenericStruct in MooseTypes.h
## Reason
`GenericStruct` was more or less just me reinventing `std::conditional` 😆 
This construct is used to simplify the creation of templates that resolve to an AD or a non-AD version of a specific type (tensors, vectors, etc.).

## Design
Use `std::conditional` plus template aliasing. Also maybe rename it to `GenericType` if that doesn't break stuff.

## Impact
Code maintenance.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #16102

### CODE REVIEW

File: modules/tensor_mechanics/include/materials/StrainEnergyRateDensity.h [hunk 1]
```diff
@@ -16,21 +16,12 @@
...
-  typedef ADRadialReturnCreepStressUpdateBase type;
-};
-}
-
+// select the appropriate class based on the is_ad boolean parameter
 template <bool is_ad>
 using GenericRadialReturnCreepStressUpdateBase =
-    typename Moose::GenericStruct<RadialReturnCreepStressUpdateBase, is_ad>::type;
+    typename std::conditional<is_ad,
+                              ADRadialReturnCreepStressUpdateBase,
```
Comment (tophmatthews) [inline modules/tensor_mechanics/include/materials/StrainEnergyRateDensity.h:23]: or....template RadialReturn?

Comment (dschwen) [reply modules/tensor_mechanics/include/materials/StrainEnergyRateDensity.h:23]: Yes, of course. This is the cleaner short term solution.

**Review by recuero — COMMENTED** [review]
Looks good to me

**Review by recuero — APPROVED** [review]

**Comment by recuero** [comment]
Two checks appear pending, but they seem to have finished properly.

**Review by lindsayad — APPROVED** [review]

**Pull request merged**